### ISSUE TITLE
Generate sublime-syntax file as well

### ISSUE THREAD

**Issue by Keats** [nim-lang/NimLime#90]
`tmLanguage` is not recommended anymore for ST3 so it would be nice to get the new format.
I tried to use the `Tools > Developer > New syntax from Nim.tmLanguage` but it fails with `Exception: unknown pattern type: dict_keys(['comment', 'name', 'patterns'])` so it doesn't look like I can just convert it automatically sadly.

**Comment by Keats** [comment]
(Just moving the conversation to this issue rather than Gutenberg)

The issue is with those bits: https://github.com/Varriount/NimLime/blob/master/Syntaxes/Nim.tmLanguage#L96

I asked on the ST Discord and @FichteFoll said the following:

> other than that, I find it hard to determine what a pattern with comment, name and patterns is supposed to do. It's not gonna match anything since there is no match, begin or end
> it can probably be dropped safely

If i remove the exception triggered by those bits, it generates the following syntax: https://gist.github.com/Keats/64c673901940f011c94e5d81e4b1b58e

**Comment by tajmone** [comment]
>  so it doesn't look like I can just convert it automatically sadly.

Conversion to the new `.sublime-syntax` format will indeed require some manual work, but I deem it an effort worth it's while. Sticking to the old `tmLanguage` format is most likely unworth the effort, for realistically there are going to be very few ST2 users left — and the new format is much better to edit, and brings better control. 

Relying on YAML to XML conversion via the `.YAML-tmLanguage` intermediate format doesn't provide the same benefits as using the new format (IMO anyway).

I've suggest in #139 adding an automated sytnax test suite, via ST3 built-in functionality. This good be a good starting point in view of porting the Nim syntax to ST3, for once the test files work correctly with the current syntax they would simplify conversion to the new format as the tests would catch any overseen problems (they would allow to ensure that the new syntax behaves as the older one).

**Comment by Keats** [comment]
I converted it some time ago: https://github.com/getzola/zola/blob/master/sublime_syntaxes/Nim.sublime-syntax

**Comment by Varriount** [comment]
The problem is that GitHub relies on the syntax files here - I'm not sure I can switch to Sublime's new syntax format.

**Comment by tajmone** [comment]
> The problem is that GitHub relies on the syntax files here...

In which sense? How does GitHub relate to them, via Travis CI you mean?

**Comment by tajmone** [comment]
> I converted it some time ago: https://github.com/getzola/zola/blob/master/sublime_syntaxes/Nim.sublime-syntax

Excellent job @Keats ! Would that be under the same MIT license of NimLime (ie, being a derivative work)? Ultimately, if there are problems switching to the new syntax in the official repository I might consider using your ported version (license permitting).

Back to the original suggestion of switching to the new ST3 syntax, I guess that dropping the old syntax in the repository might lead ST2 users to not being able to use the package any longer. In the past it was possible to add specific settings to the package.io manager so that ST2 and ST3 could "download" from different branches, but if I remember correctly this is no longer the case (but I might be wrong, I'll check).

In any case, this would mean maintaining two separate version, on different branches, rather than just one.

It would interesting to know how many ST2 users are there effectively using the extension, so that it might be possible to consider if it's worth keeping the old syntax or not. In any case, I think that sacrificing new ST3 functionality in order to accomodate ST2 support is a bit of a drawback.

**Comment by Keats** [comment]
It is the MIT licence yes, it was generated automatically by ST after all :) I only removed the bits that didn't make too much sense (in my second comment).
I don't use Nim though, so someone should probably check it!

**Comment by Varriount** [comment]
@tajmone In the sense that they are used by GitHub's highlighting feature. See https://github.com/github/linguist

**Comment by tajmone** [comment]
> @tajmone In the sense that they are used by GitHub's highlighting feature.

I didn't realize that Linguist used ST's TextMate grammars for syntax highlighting (quite cool actually).

Well, this is definitely a delicate problem which must be kept into consideration.

I see that Liguist includes the whole NimLime repo as submodule inside `vendor/grammars/`. Since a sumbodule points to a specific commit, as long as Linguist settings won't be updating the NimLime commit number it should be OK since the tmLanguage file will always be present at that commit.

But this raises a further issue, i.e. that since Nim hasn't yet reached a stable version it will still undergo syntax changes, therefore the tmLanguage file should still be updated as Nim evolves since it's used for highlighting on GitHub (we wouldn't want to loose this!).

I quickly checked the [packagecontrol docs](https://packagecontrol.io/docs) and it seems that multi-branch packages are no longer supported, which would have allowed to maintain a ST2 branch in the project.

## Having Old & New Formats Alongside?

I was wondering how ST would behave if in the same project coexisted two same-named syntax definitions, one with the old format and the other with the newer one (i.e. `Nim.tmLanguage` alongside `Nim.sublime-syntax`). ST2 should ignore the `Nim.sublime-syntax` file, since it's not aware of that format. If ST3 were to give precedence to the new format and ignore the old one than the two could live side by side and we'd see no loss of ST2 users nor Linguist support.

I should check that on ST forum, for this would be the best solution problably — even if the tmLanguage file was kept solely for the sake of GitHub Linguist (and the package was no longer aimed at ST2 support), for keeping the TextMate syntax in the project would simplify its maintainance (especially if the test suite could be run against both formats).

**Comment by tajmone** [comment]
I've asked on ST forum:

- https://forum.sublimetext.com/t/old-new-syntax-in-same-package-tmlanguage-sublime-syntax/41194

**Comment by tajmone** [comment]
Update: on Sublime Text forum I've got a reply confirming that it's ok to have both the old and new syntaxes in a same package:

- https://forum.sublimetext.com/t/old-new-syntax-in-same-package-tmlanguage-sublime-syntax/41194/2?u=tajmone

ST2 will ignore the new `Nim.sublime-syntax`, and ST3 will use only the new one and ignore `Nim.tmLanguage`.

So, it seems that at least passing to the new format won't break up GitHub Linguist. Wether it would be worthy to maintain a package fully supporting ST2 and ST3 feature might another issue altogether (in terms of python code, etc.); but it seems doable.

**Comment by FichteFoll** [comment]
FWIW, multi-branch packages are definitely supported, but they need to be done with differently prefixed tags instead of just declaring a branch to fetch updates from. That way you effectively have two separate release channels with entirely different versions.

But as you mentioned, you can just keep updating the `tmLanguage` file in parallel to the new `sublime-syntax` one and ST3 will ignore it.

**Comment by tajmone** [comment]
Thanks @FichteFoll ! As always, you are source of precious information on ST matters.

So packages via branches is now replaced by different repositories altogher? Did I understand correctly?

It's been a while since I've released my last ST package, and I remember that the PackageControl documentation mentioned something about using different branches being abrogated in the future, but when I went to check the documentation in these days I noticed that there are have been quite some changes and haven't managed to work out how this works now (not that I managed to read the whole lot etiher, just sifted rapidly).

**Comment by Varriount** [comment]
My remaining hesitation stems from the fact that keeping both files would mean that they both have to be maintained.

**Comment by FichteFoll** [comment]
>So packages via branches is now replaced by different repositories altogher? Did I understand correctly?

No, you may still use different branches in a repo, but PC won't look at them as it will only care about the tags that you push and their naming. Check out my [CSScheme package](https://github.com/FichteFoll/sublime_packages/blob/42615c564147a2337544ffd0066983a5392975c0/package_control.json#L5-L18), for example.

**Comment by AmjadHD** [comment]
#157, among other things, adds sublime-syntax files alongside the tmLanguage files.
But developing a TM grammar becomes tricky when there's a sublime-syntax with the same name, since ST prefers the latter.
So I think the way to go is to create two branches: one with TM grammars only for linguist and the other with sublime-syntax files only for Sublime Text (and Merge).

**Comment by Varriount** [comment]
@AmjadHD I'm not too sure about that - it means doing twice the work whenever the syntax needs to be updated. I'd rather just stick to the syntax format Linguist supports.

**Comment by AmjadHD** [comment]
@Varriount There's no much work to be duplicated anyway. The last 3 commits are from 2022, 2021 and 2018.
With #157 and maybe following PRs and Nim being stable there's no need to stick with a deprecated format.

**Comment by Varriount** [comment]
Sure, but even so, why do double the work? The "deprecated" format will always need to have equivalent functionality to the "new" format, which negates virtually all benefits of the new format.

**Comment by AmjadHD** [comment]
There is no need for the two syntax definitions to be equivalent feature-wise.
The new format can be more precise and faster.
Many other PLs provide sublime-syntax files, they're on package control.

**Comment by AmjadHD** [comment]
@Varriount what's your decision ?
